### PR TITLE
API Editor can publish articles, with properties

### PR DESCRIPTION
**Pull request by ViaMarcus**
[PT Card](https://www.pivotaltracker.com/story/show/172888065)
- Adding admin/articles :update
- Adding tons of tests

### CODE REVIEW

**Review by kermit-klein — APPROVED** [review]
NEAT

**Review by designerofthing — APPROVED** [review]

File: app/controllers/api/admin/articles_controller.rb [hunk 1]
```diff
@@ -9,6 +9,21 @@ def index
...
+        article.category = params[:category] || article.category
+        article.published = true
+        article.save
+        render json: { message: 'Article successfully published!'}
+      rescue => e
+        render json: { message: "Article not published: " + e.to_s }, status: 422
+      end
+    end
+  end
+
```
Comment (PaulineBA) [inline app/controllers/api/admin/articles_controller.rb:26]: What does  {+e.to_s}  mean?

Comment (ViaMarcus) [reply app/controllers/api/admin/articles_controller.rb:26]: It is the error, and it converts it to_string, so it gets the message from the error.

Comment (ViaMarcus) [reply app/controllers/api/admin/articles_controller.rb:26]: And the plus appends it to the previous string. You can see the results in the test file.

File: app/controllers/api/admin/articles_controller.rb [hunk 2]
```diff
@@ -9,6 +9,21 @@ def index
...
 
+  def update
+    if params[:activity] == "PUBLISH"
+      begin
+        article = Article.find(params[:id])
+        article.premium = params[:premium] || article.premium
+        article.category = params[:category] || article.category
+        article.published = true
+        article.save
+        render json: { message: 'Article successfully published!'}
```
Comment (erikbjoern) [inline app/controllers/api/admin/articles_controller.rb:20]: We should probably have a `if article.persisted?` here?

File: app/controllers/api/admin/articles_controller.rb (hunk 1 above)
Comment (erikbjoern) [reply app/controllers/api/admin/articles_controller.rb:26]: Personally I think string interpolation/templating makes more sense when reading it.

File: spec/requests/api/admin/editor_can_publish_article_spec.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    it 'gives a success message' do
+      expect(response_json['message']).to eq "Article successfully published!"
+    end
+
+    it 'has set published to true' do
+      expect(article.published).to eq true
+    end
+
+    it 'has set category to economy' do
+      expect(article.category).to eq "economy"
```
Comment (erikbjoern) [inline spec/requests/api/admin/editor_can_publish_article_spec.rb:32]: Where did we set it to economy?

File: spec/requests/api/admin/editor_can_publish_article_spec.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    it 'updates the provided params' do
+      expect(article.category).to eq 'economy'
+    end
+
+    it 'but not the not provided ones' do
+      expect(article.premium).to eq false
+    end
+  end
+
+  describe 'article with id does not exist' do
```
Comment (erikbjoern) [inline spec/requests/api/admin/editor_can_publish_article_spec.rb]: This is a bit confusing description, maybe could say 'when there is no article with the given id' or something like that?

**Review by erikbjoern — COMMENTED** [review]
Left a few comments, nothing big. Good work!

File: app/controllers/api/admin/articles_controller.rb (hunk 1 above)
Comment (erikbjoern) [reply app/controllers/api/admin/articles_controller.rb:26]: You could also do e.message, which maybe is a bit more readable.

Comment (oliverochman) [reply app/controllers/api/admin/articles_controller.rb:26]: https://apidock.com/rails/ActiveModel/Errors/full_messages

File: app/controllers/api/admin/articles_controller.rb (hunk 2 above)
Comment (ViaMarcus) [reply app/controllers/api/admin/articles_controller.rb:20]: We considered doing a if article.save (.persisted? is for creation I think, and will probably return true regardless of if .save worked or not), but we are inside a begin rescue, so any failure to save will be thrown as error, we are testing a lot for this. It is not needed imo to have a save check.

File: spec/requests/api/admin/editor_can_publish_article_spec.rb (hunk 3 above)
Comment (ViaMarcus) [reply spec/requests/api/admin/editor_can_publish_article_spec.rb:32]: Line 13 we send category as a param, which is economy, sorry for long line, added line breaks to puts

File: app/controllers/api/admin/articles_controller.rb (hunk 1 above)
Comment (ViaMarcus) [reply app/controllers/api/admin/articles_controller.rb:26]: Addressed this and made it more clear.

File: spec/requests/api/admin/editor_can_publish_article_spec.rb (hunk 4 above)
Comment (ViaMarcus) [reply spec/requests/api/admin/editor_can_publish_article_spec.rb]: Slightly rephrased

File: app/controllers/api/admin/articles_controller.rb (hunk 2 above)
Comment (erikbjoern) [reply app/controllers/api/admin/articles_controller.rb:20]: Ah, I see. That makes sense.

**Review by erikbjoern — APPROVED** [review]
👍

**Review by oliverochman — APPROVED** [review]

**Pull request merged**